<a href="https://colab.research.google.com/github/CodeSimple0496/No-Tutorial-AI-Data-Scientist-Roadmap-6-8-Months-of-Real-Execution/blob/main/MONTH%201/DAY%202/Data_handling_using_pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## 1. Learn Basic Pandas Operations (Their given below keyword uses to Overview the Data's Health.)

1️⃣ Dataset Overview

How many rows and columns are in the dataset?

What are the data types of each column?

Which columns are categorical and which are numerical?

In [50]:
# Import libraries
import pandas as pd

# Load Dataset
df = pd.read_csv('/content/Employee.csv')

# Find out the data Overview

print("Access the top 5 rows")
display(df.head())  # Display the first few rows(5 rows) of the DataFrame
print("Accss the last 5 rows")
display(df.tail())  # Display the last few rows(5 rows) of the DataFrame

df.info()  # Get a summary of the DataFrame
display(df.describe())  # Get statistical summary of numerical columns


# Access 273 to 489 row in the data.
display(df[273:489])

# total size of the dataset
df.shape

Access the top 5 rows


Education  JoiningYear       City  PaymentTier  Age  Gender EverBenched  \
0  Bachelors         2017  Bangalore            3   34    Male          No   
1  Bachelors         2013       Pune            1   28  Female          No   
2  Bachelors         2014  New Delhi            3   38  Female          No   
3    Masters         2016  Bangalore            3   27    Male          No   
4    Masters         2017       Pune            3   24    Male         Yes   

   ExperienceInCurrentDomain  LeaveOrNot  
0                          0           0  
1                          3           1  
2                          2           0  
3                          5           1  
4                          2           1

Accss the last 5 rows


Education  JoiningYear       City  PaymentTier  Age  Gender EverBenched  \
4648  Bachelors         2013  Bangalore            3   26  Female          No   
4649    Masters         2013       Pune            2   37    Male          No   
4650    Masters         2018  New Delhi            3   27    Male          No   
4651  Bachelors         2012  Bangalore            3   30    Male         Yes   
4652  Bachelors         2015  Bangalore            3   33    Male         Yes   

      ExperienceInCurrentDomain  LeaveOrNot  
4648                          4           0  
4649                          2           1  
4650                          5           1  
4651                          2           0  
4652                          4           0

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4653 entries, 0 to 4652
Data columns (total 9 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   Education                  4653 non-null   object
 1   JoiningYear                4653 non-null   int64 
 2   City                       4653 non-null   object
 3   PaymentTier                4653 non-null   int64 
 4   Age                        4653 non-null   int64 
 5   Gender                     4653 non-null   object
 6   EverBenched                4653 non-null   object
 7   ExperienceInCurrentDomain  4653 non-null   int64 
 8   LeaveOrNot                 4653 non-null   int64 
dtypes: int64(5), object(4)
memory usage: 327.3+ KB


JoiningYear  PaymentTier          Age  ExperienceInCurrentDomain  \
count  4653.000000  4653.000000  4653.000000                4653.000000   
mean   2015.062970     2.698259    29.393295                   2.905652   
std       1.863377     0.561435     4.826087                   1.558240   
min    2012.000000     1.000000    22.000000                   0.000000   
25%    2013.000000     3.000000    26.000000                   2.000000   
50%    2015.000000     3.000000    28.000000                   3.000000   
75%    2017.000000     3.000000    32.000000                   4.000000   
max    2018.000000     3.000000    41.000000                   7.000000   

        LeaveOrNot  
count  4653.000000  
mean      0.343864  
std       0.475047  
min       0.000000  
25%       0.000000  
50%       0.000000  
75%       1.000000  
max       1.000000

Education  JoiningYear       City  PaymentTier  Age  Gender EverBenched  \
273    Masters         2013  New Delhi            3   25    Male          No   
274  Bachelors         2018  Bangalore            3   28  Female         Yes   
275  Bachelors         2013       Pune            3   27  Female          No   
276  Bachelors         2012  Bangalore            3   25    Male          No   
277  Bachelors         2014  Bangalore            3   28    Male          No   
..         ...          ...        ...          ...  ...     ...         ...   
484  Bachelors         2012  Bangalore            3   24    Male          No   
485    Masters         2017       Pune            2   24    Male          No   
486  Bachelors         2015       Pune            2   25  Female         Yes   
487  Bachelors         2015  New Delhi            3   26  Female         Yes   
488  Bachelors         2013       Pune            2   24    Male         Yes   

     ExperienceInCurrentDomain  LeaveOrNot  
273                          3           0  
274                          1           1  
275                          5           1  
276                          3           0  
277                          1           0  
..                         ...         ...  
484                          2           1  
485                          2           0  
486                          3           1  
487                          4           0  
488                          2           1  

[216 rows x 9 columns]

(4653, 9)

- Handling Missing Data

Drop missing rows

Fill missing values

Drop the Duplicate rows

Compare before vs after

In [51]:
# 1. Drop missing rows
df = df.dropna()

# 2. FIll the missing values
df = df.fillna(0)

# 3. Drop the Duplicate rows
df = df.drop_duplicates().reset_index(drop=True)

df.shape
# After the Handling data we find out many duplicate data so it have provide more accuracy.

(2764, 9)

In [52]:
# Get the number of rows and columns
print(f"1. The dataset has {df.shape[0]} rows and {df.shape[1]} columns.")

# Get and print the data types of each column
print("\n2. Data types of each column:")
print(df.dtypes)

# Identify categorical columns (usually 'object' type in pandas)
categorical_cols = df.select_dtypes(include='object').columns.tolist()

# Identify numerical columns (usually 'int64' or 'float64' types)
numerical_cols = df.select_dtypes(include=['int64', 'float64']).columns.tolist()

# Print the identified categorical and numerical columns
print(f"\nCategorical columns: {categorical_cols}")
print(f"Numerical columns: {numerical_cols}")

1. The dataset has 2764 rows and 9 columns.

2. Data types of each column:
Education                    object
JoiningYear                   int64
City                         object
PaymentTier                   int64
Age                           int64
Gender                       object
EverBenched                  object
ExperienceInCurrentDomain     int64
LeaveOrNot                    int64
dtype: object

Categorical columns: ['Education', 'City', 'Gender', 'EverBenched']
Numerical columns: ['JoiningYear', 'PaymentTier', 'Age', 'ExperienceInCurrentDomain', 'LeaveOrNot']


2️⃣ Basic Statistics

What is the average age of employees?

What is the minimum and maximum age?

What is the average experience in the current domain?

In [53]:
# Average Age of Employees.
average_age = df['Age'].mean()
print(f"1. The average age of employees is {average_age:.2f} years.")

# Min and Max Age
min_age = df['Age'].min()
max_age = df['Age'].max()
print(f"2. The minimum age of employees is {min_age} years and the maximum age is {max_age} years.")

# Average Experience
average_exp = df['ExperienceInCurrentDomain'].mean()
print(f"3. The average experience in the current domain is {average_exp:.2f} years.")

1. The average age of employees is 30.95 years.
2. The minimum age of employees is 22 years and the maximum age is 41 years.
3. The average experience in the current domain is 2.64 years.


3️⃣ Filtering Data

Show employees older than 35.

Show employees with PaymentTier = 3.

Show female employees who left the company.

In [54]:
print("Employees older than 35:")
older_then_35 = df['Age'] > 35
display(df[older_then_35])

print("Employees with PaymentTier = 3:")
payment_tire_3 = df['PaymentTier'] == 3
display(df[payment_tire_3])

print("Female employees who left the company:")
left_female_emp = (df['Gender'] == 'Female') & (df['LeaveOrNot'] == 0)
display(df[left_female_emp])

Employees older than 35:


Education  JoiningYear       City  PaymentTier  Age  Gender EverBenched  \
2     Bachelors         2014  New Delhi            3   38  Female          No   
6     Bachelors         2015  New Delhi            3   38    Male          No   
9       Masters         2017  New Delhi            2   37    Male          No   
13    Bachelors         2016  Bangalore            3   39    Male          No   
14    Bachelors         2012  Bangalore            3   37    Male          No   
...         ...          ...        ...          ...  ...     ...         ...   
2747  Bachelors         2013       Pune            3   37    Male          No   
2749  Bachelors         2013  Bangalore            3   39  Female          No   
2752  Bachelors         2016  New Delhi            1   38  Female          No   
2753    Masters         2014  New Delhi            3   37  Female          No   
2761    Masters         2013       Pune            2   37    Male          No   

      ExperienceInCurrentDomain  LeaveOrNot  
2                             2           0  
6                             0           0  
9                             2           0  
13                            2           0  
14                            4           0  
...                         ...         ...  
2747                          3           0  
2749                          5           0  
2752                          2           1  
2753                          0           1  
2761                          2           1  

[666 rows x 9 columns]

Employees with PaymentTier = 3:


Education  JoiningYear       City  PaymentTier  Age  Gender EverBenched  \
0     Bachelors         2017  Bangalore            3   34    Male          No   
2     Bachelors         2014  New Delhi            3   38  Female          No   
3       Masters         2016  Bangalore            3   27    Male          No   
4       Masters         2017       Pune            3   24    Male         Yes   
5     Bachelors         2016  Bangalore            3   22    Male          No   
...         ...          ...        ...          ...  ...     ...         ...   
2757  Bachelors         2013  Bangalore            3   31  Female          No   
2758  Bachelors         2015       Pune            3   32  Female         Yes   
2760  Bachelors         2016       Pune            3   30    Male          No   
2762    Masters         2018  New Delhi            3   27    Male          No   
2763  Bachelors         2012  Bangalore            3   30    Male         Yes   

      ExperienceInCurrentDomain  LeaveOrNot  
0                             0           0  
2                             2           0  
3                             5           1  
4                             2           1  
5                             0           0  
...                         ...         ...  
2757                          5           0  
2758                          1           1  
2760                          2           0  
2762                          5           1  
2763                          2           0  

[1976 rows x 9 columns]

Female employees who left the company:


Education  JoiningYear       City  PaymentTier  Age  Gender EverBenched  \
2     Bachelors         2014  New Delhi            3   38  Female          No   
16    Bachelors         2014  Bangalore            3   34  Female          No   
25    Bachelors         2013  Bangalore            3   22  Female         Yes   
29      Masters         2017  New Delhi            2   30  Female          No   
35    Bachelors         2014  Bangalore            1   30  Female          No   
...         ...          ...        ...          ...  ...     ...         ...   
2741  Bachelors         2012  Bangalore            3   39  Female          No   
2748        PHD         2012  Bangalore            3   30  Female          No   
2749  Bachelors         2013  Bangalore            3   39  Female          No   
2757  Bachelors         2013  Bangalore            3   31  Female          No   
2759    Masters         2017       Pune            2   31  Female          No   

      ExperienceInCurrentDomain  LeaveOrNot  
2                             2           0  
16                            2           0  
25                            0           0  
29                            2           0  
35                            3           0  
...                         ...         ...  
2741                          1           0  
2748                          4           0  
2749                          5           0  
2757                          5           0  
2759                          2           0  

[621 rows x 9 columns]

4️⃣ Sorting Data

Sort employees by Age (descending).

Sort employees by ExperienceInCurrentDomain and Age together.

In [55]:
# Sort Employees Age By descending order.
print("Sort Employees Age By descending order.")
sorted_age = df.sort_values(by='Age', ascending=False)
display(sorted_age)


# Sort employees Experience and Age Together.
print('Sort employees by ExperienceInCurrentDomain and Age together')
sort_Exp_and_Age = df.sort_values(by = ['ExperienceInCurrentDomain', 'Age'], ascending=False)
display(sort_Exp_and_Age)

Sort Employees Age By descending order.


Education  JoiningYear       City  PaymentTier  Age  Gender EverBenched  \
2290  Bachelors         2015  Bangalore            3   41    Male          No   
2285        PHD         2017       Pune            3   41    Male          No   
2145  Bachelors         2017  Bangalore            3   41  Female          No   
2166    Masters         2017  Bangalore            2   41    Male         Yes   
2164  Bachelors         2015  Bangalore            3   41    Male          No   
...         ...          ...        ...          ...  ...     ...         ...   
2356  Bachelors         2017       Pune            2   22    Male          No   
91    Bachelors         2015  New Delhi            3   22    Male          No   
2358  Bachelors         2012  Bangalore            3   22    Male          No   
5     Bachelors         2016  Bangalore            3   22    Male          No   
19    Bachelors         2016  New Delhi            2   22  Female          No   

      ExperienceInCurrentDomain  LeaveOrNot  
2290                          4           0  
2285                          2           0  
2145                          2           0  
2166                          2           1  
2164                          1           0  
...                         ...         ...  
2356                          0           0  
91                            0           0  
2358                          0           0  
5                             0           0  
19                            0           1  

[2764 rows x 9 columns]

Sort employees by ExperienceInCurrentDomain and Age together


Education  JoiningYear       City  PaymentTier  Age  Gender EverBenched  \
2354  Bachelors         2015       Pune            2   41  Female          No   
2336    Masters         2014  Bangalore            3   40  Female          No   
2325  Bachelors         2016  Bangalore            3   39    Male          No   
2340  Bachelors         2014  Bangalore            3   39    Male          No   
2327  Bachelors         2016       Pune            3   38  Female          No   
...         ...          ...        ...          ...  ...     ...         ...   
2604  Bachelors         2014  Bangalore            3   22  Female          No   
2611  Bachelors         2017       Pune            2   22  Female          No   
2658  Bachelors         2017  New Delhi            2   22    Male          No   
2668  Bachelors         2015  Bangalore            3   22    Male          No   
2693  Bachelors         2012  Bangalore            3   22    Male          No   

      ExperienceInCurrentDomain  LeaveOrNot  
2354                          7           0  
2336                          7           1  
2325                          7           0  
2340                          7           1  
2327                          7           0  
...                         ...         ...  
2604                          0           0  
2611                          0           1  
2658                          0           1  
2668                          0           1  
2693                          0           1  

[2764 rows x 9 columns]

⚙️ Level 2 — Business-Oriented Pandas Questions

5️⃣ GroupBy Analysis

Average age by gender.

Count employees in each city.

Average PaymentTier by education level.

Attrition count by city (LeaveOrNot).

In [56]:
# Average age by gender
Avg_age_by_gender = df.groupby('Gender')['Age'].mean()
print("Average age by gender:")
display(Avg_age_by_gender)

# Count employees in each city
emp_count_by_city = df['City'].value_counts()
print("\nCount employees in each city:")
display(emp_count_by_city)

# Average paymentTire by education level
avg_PT_by_edu = df.groupby('Education')['PaymentTier'].mean()
print("\nAverage paymentTire by education level:")
display(avg_PT_by_edu)

# Attrition Count by City (LeaveOrNot)
attrition_count_by_city = df.groupby(['City', 'LeaveOrNot']).size()
print("\nAttrition count by city (LeaveOrNot):")
display(attrition_count_by_city)


Average age by gender:


Gender
Female    30.737652
Male      31.126880
Name: Age, dtype: float64


Count employees in each city:


City
Bangalore    1171
Pune          801
New Delhi     792
Name: count, dtype: int64


Average paymentTire by education level:


Education
Bachelors    2.677321
Masters      2.492936
PHD          2.698718
Name: PaymentTier, dtype: float64


Attrition count by city (LeaveOrNot):


City       LeaveOrNot
Bangalore  0             761
           1             410
New Delhi  0             522
           1             270
Pune       0             393
           1             408
dtype: int64

6️⃣ Attrition Analysis

What percentage of employees left the company?

Which gender has higher attrition?

Which education level has highest attrition?

Does PaymentTier affect leaving?

In [57]:
# Percentage of the employee left the company
per_emp_left = (df['LeaveOrNot'] == 1).mean() * 100
print(f"1. The percentage of employees who left the company is {per_emp_left}")

# Which gender has higher attrition
gender_attrition = df.groupby('Gender')['LeaveOrNot'].mean()
print("\n2. Attrition rate by gender:")
display(gender_attrition)

# Which Education level has highest attrition
edu_attrition = df.groupby('Education')['LeaveOrNot'].mean()
print("\n3. Attrition rate by Education:")
display(edu_attrition)

#Does PaymentTire affect leaving?
payment_attrition = df.groupby('PaymentTier')['LeaveOrNot'].mean()
print("\n4. Attrition rate by PaymentTier:")
display(payment_attrition)

## Yes, Payment is affect the Leaving the employees in Company. for the advancing the upper position.(Specially Tire-2)


1. The percentage of employees who left the company is 39.36324167872648

2. Attrition rate by gender:


Gender
Female    0.497166
Male      0.310007
Name: LeaveOrNot, dtype: float64


3. Attrition rate by Education:


Education
Bachelors    0.374937
Masters      0.485086
PHD          0.256410
Name: LeaveOrNot, dtype: float64


4. Attrition rate by PaymentTier:


PaymentTier
1    0.353211
2    0.601754
3    0.338057
Name: LeaveOrNot, dtype: float64

7️⃣ Experience Insights

Average experience of employees who left vs stayed.

Does low experience correlate with leaving?

In [58]:
# Average experience of employees who left vs stayed.
avg_exp_by_leave = df.groupby('LeaveOrNot')['ExperienceInCurrentDomain'].mean()
print("Average experience of employees who left vs stayed:")
display(avg_exp_by_leave)


# Does low experience correlate with leaving?
exp_corr_leav = df.groupby('ExperienceInCurrentDomain')['LeaveOrNot'].mean()
print("Does low experience correlate with leaving:")
display(exp_corr_leav)


Average experience of employees who left vs stayed:


LeaveOrNot
0    2.671838
1    2.602022
Name: ExperienceInCurrentDomain, dtype: float64

Does low experience correlate with leaving:


ExperienceInCurrentDomain
0    0.379791
1    0.369515
2    0.427313
3    0.434590
4    0.392941
5    0.340426
6    0.250000
7    0.333333
Name: LeaveOrNot, dtype: float64

🔥 Level 3 — Thinking Questions (Write in Markdown)

1. Which features are most important for predicting attrition?

2. Which features look useless?

3. What biases might exist in this dataset?

4. What business decisions can HR make?

Answer -

## 🔥 Level 3 — Thinking Questions (Write in Markdown)

### Answer -

1.  **Which features are most important for predicting attrition?**
    Based on the analysis, the most important features for predicting attrition appear to be:
    *   **PaymentTier**: Employees in PaymentTier 2 have a significantly higher attrition rate (60.18%).
    *   **Gender**: Female employees have a considerably higher attrition rate (49.72%) than male employees.
    *   **Education**: Masters-level employees show the highest attrition rate (48.51%).
    *   **City**: Pune shows a higher proportion of employees leaving compared to Bangalore and New Delhi.
    *   **ExperienceInCurrentDomain**: While not as strong, experience levels around 2-3 years show a slightly elevated attrition risk.

2.  **Which features look useless?**

    From the provided analysis, all features have been used or shown some correlation with attrition, except for `EverBenched`. Without specific analysis on `EverBenched` and `JoiningYear` (beyond general statistics), it's difficult to conclusively label them as 'useless' without further investigation into their correlation with `LeaveOrNot`.

3.  **What biases might exist in this dataset?**
    Several potential biases could exist:
    *   **Selection Bias**: The dataset might only represent employees from a specific company or industry, limiting generalizability to a broader population.
    *   **Reporting Bias**: Data might be collected in a way that under- or over-represents certain groups or outcomes.
    *   **Survivorship Bias**: The dataset only includes employees up to a certain point, potentially missing insights from those who left very early or stayed for a very long time if not captured accurately.
    *   **Data Cleaning Impact**: The process of dropping duplicates (from 4653 to 2764 rows) could introduce bias if the duplicates were not random or if there was a systematic reason for their presence that is not accounted for.
    *   **Gender Imbalance**: If one gender is disproportionately represented, it could skew analysis related to gender.

4.  **What business decisions can HR make?**
    Based on these insights, HR can make several data-driven decisions:
    *   **Targeted Retention Programs**: Implement specific retention strategies for employees in PaymentTier 2, particularly focusing on those with a Masters education and female employees, as these groups show the highest attrition rates.
    *   **Investigate Root Causes**: Conduct exit interviews or surveys to understand the specific reasons why employees in PaymentTier 2, Masters-level, and female employees are leaving the company at higher rates.
    *   **Geographic Focus**: Analyze workplace culture, management practices, and compensation in Pune to address the higher attrition observed there.
    *   **Career Development**: Develop career progression plans and mentorship programs for employees with 2-3 years of experience to reduce attrition during this critical period.
    *   **Review Compensation & Benefits**: Evaluate if compensation or benefits are competitive for PaymentTier 2 roles, as this is a strong indicator of dissatisfaction.
    *   **Diversity & Inclusion Initiatives**: Review D&I policies to ensure an equitable and supportive environment, especially given the higher attrition among female employees.



## Answers to the Questions

### 1️⃣ Dataset Overview

1.  **How many rows and columns are in the dataset?**
    The dataset has `2764` rows and `9` columns after handling missing values and duplicates.

2.  **What are the data types of each column?**
    *   `Education`: `object`
    *   `JoiningYear`: `int64`
    *   `City`: `object`
    *   `PaymentTier`: `int64`
    *   `Age`: `int64`
    *   `Gender`: `object`
    *   `EverBenched`: `object`
    *   `ExperienceInCurrentDomain`: `int64`
    *   `LeaveOrNot`: `int64`

3.  **Which columns are categorical and which are numerical?**
    *   **Categorical columns**: `['Education', 'City', 'Gender', 'EverBenched']`
    *   **Numerical columns**: `['JoiningYear', 'PaymentTier', 'Age', 'ExperienceInCurrentDomain', 'LeaveOrNot']`

### 2️⃣ Basic Statistics

1.  **What is the average age of employees?**
    The average age of employees is `30.95` years.

2.  **What is the minimum and maximum age?**
    The minimum age of employees is `22` years and the maximum age is `41` years.

3.  **What is the average experience in the current domain?**
    The average experience in the current domain is `2.64` years.


### 3️⃣ Filtering Data

In [ ]:
# Calculate counts for filtering questions
num_older_than_35 = df[df['Age'] > 35].shape[0]
num_payment_tier_3 = df[df['PaymentTier'] == 3].shape[0]
num_left_female_emp = df[(df['Gender'] == 'Female') & (df['LeaveOrNot'] == 1)].shape[0] # Corrected for 'left the company' (LeaveOrNot == 1)

print(f"1. Number of employees older than 35: {num_older_than_35}")
print(f"2. Number of employees with PaymentTier = 3: {num_payment_tier_3}")
print(f"3. Number of female employees who left the company: {num_left_female_emp}")

### 4️⃣ Sorting Data

1.  **Sort employees by Age (descending).**
    The employees were sorted by Age in descending order and displayed.

2.  **Sort employees by ExperienceInCurrentDomain and Age together.**
    The employees were sorted by 'ExperienceInCurrentDomain' and 'Age' together and displayed.

### 5️⃣ GroupBy Analysis

1.  **Average age by gender.**
    *   Female: `30.74` years
    *   Male: `31.13` years

2.  **Count employees in each city.**
    *   Bangalore: `1171` employees
    *   Pune: `801` employees
    *   New Delhi: `792` employees

3.  **Average PaymentTier by education level.**
    *   Bachelors: `2.68`
    *   Masters: `2.49`
    *   PHD: `2.70`

4.  **Attrition count by city (LeaveOrNot).**
    *   **Bangalore**:
        *   Stayed (`0`): `761`
        *   Left (`1`): `410`
    *   **New Delhi**:
        *   Stayed (`0`): `522`
        *   Left (`1`): `270`
    *   **Pune**:
        *   Stayed (`0`): `393`
        *   Left (`1`): `408`

### 6️⃣ Attrition Analysis

1.  **What percentage of employees left the company?**
    `39.36%` of employees left the company.

2.  **Which gender has higher attrition?**
    Female employees have a higher attrition rate (`49.72%`) compared to Male employees (`31.00%`).

3.  **Which education level has highest attrition?**
    Employees with a `Masters` degree have the highest attrition rate (`48.51%`).

4.  **Does PaymentTier affect leaving?**
    Yes, PaymentTier significantly affects leaving. PaymentTier 2 has a substantially higher attrition rate (`60.18%`) compared to PaymentTier 1 (`35.32%`) and PaymentTier 3 (`33.81%`).

### 7️⃣ Experience Insights

1.  **Average experience of employees who left vs stayed.**
    *   Employees who stayed (`LeaveOrNot = 0`): `2.67` years
    *   Employees who left (`LeaveOrNot = 1`): `2.60` years

2.  **Does low experience correlate with leaving?**
    While the average experience for those who left is slightly lower, looking at the breakdown by `ExperienceInCurrentDomain`, employees with 2 and 3 years of experience show slightly higher attrition rates (0.427 and 0.435 respectively) compared to those with 0, 1, or 4 years of experience. This suggests a nuanced correlation, where *mid-level low experience* (2-3 years) might be a critical period for attrition.


## 🔥 Level 3 — Thinking Questions (Write in Markdown)

### Answer -

1.  **Which features are most important for predicting attrition?**
    Based on the analysis, the most important features for predicting attrition appear to be:
    *   **PaymentTier**: Employees in PaymentTier 2 have a significantly higher attrition rate (60.18%).
    *   **Gender**: Female employees have a considerably higher attrition rate (49.72%) than male employees.
    *   **Education**: Masters-level employees show the highest attrition rate (48.51%).
    *   **City**: Pune shows a higher proportion of employees leaving compared to Bangalore and New Delhi.
    *   **ExperienceInCurrentDomain**: While not as strong, experience levels around 2-3 years show a slightly elevated attrition risk.

2.  **Which features look useless?**

    From the provided analysis, all features have been used or shown some correlation with attrition, except for `EverBenched`. Without specific analysis on `EverBenched` and `JoiningYear` (beyond general statistics), it's difficult to conclusively label them as 'useless' without further investigation into their correlation with `LeaveOrNot`.

3.  **What biases might exist in this dataset?**
    Several potential biases could exist:
    *   **Selection Bias**: The dataset might only represent employees from a specific company or industry, limiting generalizability to a broader population.
    *   **Reporting Bias**: Data might be collected in a way that under- or over-represents certain groups or outcomes.
    *   **Survivorship Bias**: The dataset only includes employees up to a certain point, potentially missing insights from those who left very early or stayed for a very long time if not captured accurately.
    *   **Data Cleaning Impact**: The process of dropping duplicates (from 4653 to 2764 rows) could introduce bias if the duplicates were not random or if there was a systematic reason for their presence that is not accounted for.
    *   **Gender Imbalance**: If one gender is disproportionately represented, it could skew analysis related to gender.

4.  **What business decisions can HR make?**
    Based on these insights, HR can make several data-driven decisions:
    *   **Targeted Retention Programs**: Implement specific retention strategies for employees in PaymentTier 2, particularly focusing on those with a Masters education and female employees, as these groups show the highest attrition rates.
    *   **Investigate Root Causes**: Conduct exit interviews or surveys to understand the specific reasons why employees in PaymentTier 2, Masters-level, and female employees are leaving the company at higher rates.
    *   **Geographic Focus**: Analyze workplace culture, management practices, and compensation in Pune to address the higher attrition observed there.
    *   **Career Development**: Develop career progression plans and mentorship programs for employees with 2-3 years of experience to reduce attrition during this critical period.
    *   **Review Compensation & Benefits**: Evaluate if compensation or benefits are competitive for PaymentTier 2 roles, as this is a strong indicator of dissatisfaction.
    *   **Diversity & Inclusion Initiatives**: Review D&I policies to ensure an equitable and supportive environment, especially given the higher attrition among female employees.



In [59]:
def attrition_rate(df):
    return df["LeaveOrNot"].mean() * 100

attrition_rate(df)

np.float64(39.36324167872648)

##